README.md
#Install repo
To install the repository please download the [repository.shali-1.0.0.zip](https://github.com/shali3/xbmc-repo/raw/master/repository.shali-1.0.0.zip) file and install from zip in xbmc.

#List of Addons
This repository currently have the following addons

## Red Alert Israel
This addon is under **Services** and was written to give alerts on red alerts in Israel.

After installing please enter your region code in the settings. You can get region code from the [Pikud Haoref website](http://www.oref.org.il/1096-he/Pakar.aspx). For example: Tel-Aviv is 157 and Givataim is 160.

### Features
1. Sound a loud alarm when there's a red alert in your region (can be disabled from settings)
2. Pause playback when there's an alarm in your region (on by default, can be disabled from settings)
3. Pause Live TV when there's an alarm in you region (off by default, can be disabled from settings)
 


repository.shali/changelog.txt
[B]Version 1.0.0[/B]
- Created the Repo

addons_xml_generator.py
# !/usr/bin/env python
# *
# *  Copyright (C) 2012-2013 Garrett Brown
# *  Copyright (C) 2010      j48antialias
# *
# *  This Program is free software; you can redistribute it and/or modify
# *  it under the terms of the GNU General Public License as published by
# *  the Free Software Foundation; either version 2, or (at your option)
# *  any later version.
# *
# *  This Program is distributed in the hope that it will be useful,
# *  but WITHOUT ANY WARRANTY; without even the implied warranty of
# *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# *  GNU General Public License for more details.
# *
# *  You should have received a copy of the GNU General Public License
# *  along with XBMC; see the file COPYING.  If not, write to
# *  the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
# *  http://www.gnu.org/copyleft/gpl.html
# *
# *  Based on code by j48antialias:
# *  https://anarchintosh-projects.googlecode.com/files/addons_xml_generator.py

""" addons.xml generator """

import os
import sys
import shutil
import zipfile
import xml.etree.ElementTree as ET

# Compatibility with 3.0, 3.1 and 3.2 not supporting u"" literals
if sys.version < '3':
    import codecs

    def u(x):
        return codecs.unicode_escape_decode(x)[0]
else:
    def u(x):
        return x


class Generator:
    """
        Generates a new addons.xml file from each addons addon.xml file
        and a new addons.xml.md5 hash file. Must be run from the root of
        the checked-out repo. Only handles single depth folder structure.
    """

    def __init__(self):
        # generate files
        self.excludedFolders = ['.svn', '.git', 'repo']
        self.files_to_copy = ['icon.png', 'changelog.txt', 'fanart.jpg']
        self.excludedFiles = ['.DS_Store']
        self._generate_zip_files()
        self._generate_addons_file()
        self._generate_md5_file()
        # notify user
        print("Finished updating addons xml and md5 files")

    def _generate_zip_files(self):
        # addon list
        addons = os.listdir(".")

        # loop thru and add each addons addon.xml file
        for addon in addons:
            try:
                # skip any file or .svn folder or .git folder
                if not os.path.isdir(addon) or addon in self.excludedFolders: continue

                _path = os.path.join(addon, "addon.xml")
                tree = ET.parse(_path)
                root = tree.getroot()
                zip_file_name = '%s-%s.zip' % (root.attrib['id'], root.attrib['version'])

                if not addon.startswith("repository"):  # this will zip the repositories in the root folder.
                    dir = os.path.join('repo', addon)
                    if not os.path.exists(dir):
                        os.makedirs(dir)
                    for file in self.files_to_copy:
                        file_path = os.path.join(addon, file)
                        if os.path.exists(file_path):
                            shutil.copyfile(file_path, os.path.join(dir, file))
                            print "coping " + file + " for " + addon
                    zip_file_name = os.path.join(dir, zip_file_name)

                zip = zipfile.ZipFile(zip_file_name, 'w')
                for root, dirs, files in os.walk(addon):
                    for file in files:
                        if files not in self.excludedFiles:
                            zip.write(os.path.join(root, file))
                zip.close()
                print 'zipped ' + zip_file_name
            except Exception as e:
                # missing or poorly formatted addon.xml
                print("Excluding %s for %s" % ( addon, e ))


    def _generate_addons_file(self):
        # addon list
        addons = os.listdir(".")

        # final addons text
        addons_xml = u("<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"yes\"?>\n<addons>\n")
        # loop thru and add each addons addon.xml file
        for addon in addons:
            try:
                # skip any file or .svn folder or .git folder
                if not os.path.isdir(addon) or addon in self.excludedFolders or addon.startswith("repository"): continue
                # create path
                _path = os.path.join(addon, "addon.xml")
                # split lines for stripping
                xml_lines = open(_path, "r").read().splitlines()
                # new addon
                addon_xml = ""
                # loop thru cleaning each line
                for line in xml_lines:
                    # skip encoding format line
                    if ( line.find("<?xml") >= 0 ): continue
                    # add line
                    if sys.version < '3':
                        addon_xml += unicode(line.rstrip() + "\n", "UTF-8")
                    else:
                        addon_xml += line.rstrip() + "\n"
                # we succeeded so add to our final addons.xml text
                addons_xml += addon_xml.rstrip() + "\n\n"
            except Exception as e:
                # missing or poorly formatted addon.xml
                print("Excluding %s for %s" % ( _path, e ))
        # clean and add closing tag
        addons_xml = addons_xml.strip() + u("\n</addons>\n")
        # save file
        self._save_file(addons_xml.encode("UTF-8"), file="addons.xml")

    def _generate_md5_file(self):
        # create a new md5 hash
        try:
            import md5

            m = md5.new(open("addons.xml", "r").read()).hexdigest()
        except ImportError:
            import hashlib

            m = hashlib.md5(open("addons.xml", "r", encoding="UTF-8").read().encode("UTF-8")).hexdigest()

        # save file
        try:
            self._save_file(m.encode("UTF-8"), file="addons.xml.md5")
        except Exception as e:
            # oops
            print("An error occurred creating addons.xml.md5 file!\n%s" % e)

    def _save_file(self, data, file):
        try:
            # write data to the file (use b for Python 3)
            open(file, "wb").write(data)
        except Exception as e:
            # oops
            print("An error occurred saving %s file!\n%s" % ( file, e ))


if ( __name__ == "__main__" ):
    # start
    Generator()

script.service.redalert/main.py
import xbmc
import xbmcaddon
import xbmcgui
import threading
import time
import json
import urllib2
import re
import os
import datetime
import sys

__addon__ = xbmcaddon.Addon()
__cwd__ = __addon__.getAddonInfo('path')
__scriptname__ = __addon__.getAddonInfo('name')
__version__ = __addon__.getAddonInfo('version')
__icon__ = __addon__.getAddonInfo('icon')
__ID__ = __addon__.getAddonInfo('id')
__language__ = __addon__.getLocalizedString

# LIB_PATH = xbmc.translatePath( os.path.join( __cwd__, 'resources', 'lib' ) )
# sys.path.append (LIB_PATH)

sound_file_path = xbmc.translatePath(os.path.join(__cwd__, 'resources', 'alert.wav'))
cities_file_path = xbmc.translatePath(os.path.join(__cwd__, 'resources', 'cities.json'))
json_data = open(cities_file_path)
area = json.load(json_data)
json_data.close()


def log(msg):
    try:
        xbmc.log(u"### [%s] - %s" % (__scriptname__, msg,), level=xbmc.LOGDEBUG)
    except:
        pass


def play_wav(full_path):
    try:
        platform = sys.platform
        if platform == "darwin":
            import subprocess

            subprocess.call(["afplay", full_path])
        elif platform == "win32":
            import winsound

            winsound.PlaySound(full_path, winsound.SND_FILENAME)
        elif platform == "linux" or platform == "linux2":
            from wave import open as waveOpen
            from ossaudiodev import open as ossOpen

            s = waveOpen('tada.wav', 'rb')
            (nc, sw, fr, nf, comptype, compname) = s.getparams()
            dsp = ossOpen('/dev/dsp', 'w')
            try:
                from ossaudiodev import AFMT_S16_NE
            except ImportError:
                if byteorder == "little":
                    AFMT_S16_NE = ossaudiodev.AFMT_S16_LE
                else:
                    AFMT_S16_NE = ossaudiodev.AFMT_S16_BE
            dsp.setparameters(AFMT_S16_NE, nc, fr)
            data = s.readframes(nf)
            s.close()
            dsp.write(data)
            dsp.close()
    except:
        log('Failed to play file %s' % full_path);
        if getSettingAsBool('play_alert_fallback'):
            xbmc.Player().play(full_path)


log("[%s] - Version: %s Started" % (__scriptname__, __version__))

# helper function to get string type from settings
def getSetting(setting):
    return __addon__.getSetting(setting).strip()


# helper function to get bool type from settings
def getSettingAsBool(setting):
    return getSetting(setting).lower() == "true"


last_id = 0


def fetch_data(url, utf16=True):
    try:
        response = urllib2.urlopen(url)
        if response.code == 200:
            text = response.read()
            if utf16:
                text = unicode(text, "utf-16")
            obj = json.loads(text)
            return obj
        else:
            log('failed getting data from %s CODE %d' % (url, response.code))
    except:
        log('failed getting data from %s' % url)


def notify_alert(cities, in_user_region=False):
    all_cities = [item for sublist in cities.values() for item in sublist]
    time = str(datetime.datetime.now().time())
    time = time[:time.find('.')]

    play_alert = in_user_region and getSettingAsBool('play_alert')

    timeout = 5000
    if in_user_region:
        timeout = 90000
    xbmcgui.Dialog().notification("Red Alert %s" % time,
                                  "%s (%s)" % (','.join(all_cities), ','.join(cities.keys()))
                                  , __icon__, timeout, not play_alert)
    player = xbmc.Player()
    if player.isPlaying():
        live_tv = player.getPlayingFile().find("pvr://") > -1
        if (live_tv and getSettingAsBool('pause_live_tv')) or (not live_tv and getSettingAsBool('pause_other')):
            player.pause()

    if play_alert:
        play_wav(sound_file_path)


def process_json(obj):
    global last_id
    if obj:
        new_id = obj.get("id")
        if new_id != last_id:
            alert_regions = re.findall("\d+", getSetting('alert_code'));
            alert_cities = []
            last_id = new_id
            cities = {}  # code to cities names
            data = obj.get("data")

            for item in data:
                for city in item.split(','):
                    city_name = re.sub("\d+", "", city).strip()
                    code = re.sub("[^\d]+", "", city).strip()

                    if len(alert_regions) == 0 or code in alert_regions:
                        city_codes = cities.get(code, [])
                        city_codes += area.get(city, [])
                        cities[code] = city_codes

            if len(cities) > 0:
                notify_alert(cities, len(alert_regions) > 0)


def check_for_alerts():
    global last_id
    # obj = fetch_data('http://dl.dropboxusercontent.com/u/8058232/testing/alerts.json', False);  # Testing url...
    obj = fetch_data('http://www.oref.org.il/WarningMessages/alerts.json', True);
    process_json(obj)


if len(getSetting('alert_code').strip()) == 0:
    response = xbmcgui.Dialog().ok("Red Alert Israel",
                                   'You need to specify region codes you want to be alerted on. '
                                   'You can find them at oref.gov.il.\n'
                                   'You can specify more than one by separating with spaces.\n'
                                   'For example: Tel-Aviv is 157 and Givataim is 160.')
    if response:
        xbmcaddon.Addon().openSettings()

counter = 0;
while not xbmc.abortRequested:
    if counter == 0:
        check_for_alerts()
    counter = (counter + 1) % 30
    xbmc.sleep(100)  # low sleep to keep the quit process quick